opendevin/controller/__init__.py

```diff
@@ -0,0 +1,90 @@
+
+from typing import List, Callable
+
+from opendevin.state import State
+from opendevin.agent import Agent
+from opendevin.action import (
+    Action,
+    FileReadAction,
+    FileWriteAction,
+    AgentFinishAction,
+)
+from opendevin.observation import (
+    Observation,
+)
+
+
+from .command_manager import CommandManager
+
+
+def print_callback(event):
+    print(event, flush=True)
+
+
+class AgentController:
+    def __init__(
+        self,
+        agent: Agent,
+        workdir: str,
+        max_iterations: int = 100,
+        callbacks: List[Callable] = [],
+    ):
+        self.agent = agent
+        self.max_iterations = max_iterations
+        self.workdir = workdir
+        self.command_manager = CommandManager(workdir)
+        self.state_updated_info: List[Action | Observation] = []
```

Comment:
I guess one issue we are having right now is that user can actually send over their agent to the control loop, which might make `Tuple[Action, Observation]` challenge, should agent always treat that as an observation? Or do we always enforce that only thing comes from the front-end can only be `Observation` (which will simplify things by a ton!!)

Reply:
Hmm...interesting question.

I'd actually say that the user input should always be an `Action`. Messages could be something like the `think` action we currently have. Maybe `hint`?

But I imagine the user could also `run` a command to help the agent (e.g. maybe it's struggling to figure out `apt-get install npm`)

Reply:
FWIW there's some video with Devin struggling with a task, and the user saying "wait, don't do that, do <something else>".

Reply:
as an aside, I think we'll want to have an API endpoint that returns the history, so the user can see everything happening (at least for debugging purposes).

Giving the items in the history a consistent format will help a bunch there

Reply:
From a research perspective, I guess allowing the user to execute a command in Devin's execution environment naturally makes the agent's environment change from [static to dynamic](https://www.javatpoint.com/agent-environment-in-ai), which might pose a bunch of challenges in both implementations and research. To the best of my knowledge, most agent research today mostly assumes a static environment: [very few agent benchmarks are tailored for temporal changes](https://arxiv.org/abs/2402.07744), let alone a dynamic environment.

So I think maybe it will be beneficial for us to focus on the dynamic environment first (user just watch and talk, but do not execute actions themselves). Maybe after we get everything work smoothly and our agents can solve SWE-Bench issues, then we can start studying this problem with dynamic environment, how do y'all think?

PLUS: i think it will be very nice to keep track of history using Controller, we can starts a new issue to call for help after this PR is merged.